lib/persist-state.c

```diff
@@ -174,6 +175,22 @@ _wait_until_map_release(PersistState* self)
 }
 
 static gboolean
+_increase_file_size(PersistState *self, guint32 new_size)
+{
+  ssize_t length = new_size - self->current_size;
+  gchar *pad_buffer = g_new0(gchar, length);
+  gboolean result = pwrite_strict(self->fd, pad_buffer, length, self->current_size);
+  if (!result)
+    {
+      msg_error("Can't grow the persist file", evt_tag_int("old_size", self->current_size), evt_tag_int("new_size", new_size), evt_tag_errno("error", errno), NULL);
+    }
+  g_free(pad_buffer);
+  return result;
+}
```

Comment:
Thanks algernon!

I could not have described it better! :)

On Fri, Sep 25, 2015 at 2:51 PM, Gergely Nagy <notifications@github.com>
wrote:

> In lib/persist-state.c
> <https://github.com/balabit/syslog-ng/pull/717#discussion_r40425646>:
>
> > @@ -174,6 +175,22 @@ _wait_until_map_release(PersistState* self)
> >  }
> >
> >  static gboolean
> > +_increase_file_size(PersistState *self, guint32 new_size)
> > +{
> > +  ssize_t length = new_size - self->current_size;
> > +  gchar *pad_buffer = g_new0(gchar, length);
> > +  gboolean result = pwrite_strict(self->fd, pad_buffer, length, self->current_size);
> > +  if (!result)
> > +    {
> > +      msg_error("Can't grow the persist file", evt_tag_int("old_size", self->current_size), evt_tag_int("new_size", new_size), evt_tag_errno("error", errno), NULL);
> > +    }
> > +  g_free(pad_buffer);
> > +  return result;
> > +}
>
> It's not enough because of sparse files. It works, but for example on
> ext4, that'll create a sparse file, which means that if you have 3k free
> space, you seek to 4k and write a byte, that succeeds. But as soon as you
> try to write past the ~3k mark, you'll get a SIGBUS (due to mmap, if you'd
> write, you'd get an error, not sure which one).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/717/files#r40425646>.
>